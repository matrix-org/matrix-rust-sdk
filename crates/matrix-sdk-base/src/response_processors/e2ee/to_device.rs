// Copyright 2025 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::BTreeMap;

use matrix_sdk_crypto::{store::RoomKeyInfo, EncryptionSyncChanges, OlmMachine};
use ruma::{
    api::client::sync::sync_events::{v3, v5, DeviceLists},
    events::AnyToDeviceEvent,
    serde::Raw,
    OneTimeKeyAlgorithm, UInt,
};

use crate::Result;

/// Process the to-device events and other related e2ee data based on a response
/// from a [MSC4186 request][`v5`].
///
/// This returns a list of all the to-device events that were passed in but
/// encrypted ones were replaced with their decrypted version.
pub async fn from_msc4186(
    to_device: Option<&v5::response::ToDevice>,
    e2ee: &v5::response::E2EE,
    olm_machine: Option<&OlmMachine>,
) -> Result<Output> {
    process(
        olm_machine,
        to_device.as_ref().map(|to_device| to_device.events.clone()).unwrap_or_default(),
        &e2ee.device_lists,
        &e2ee.device_one_time_keys_count,
        e2ee.device_unused_fallback_key_types.as_deref(),
        to_device.as_ref().map(|to_device| to_device.next_batch.clone()),
    )
    .await
}

/// Process the to-device events and other related e2ee data based on a response
/// from a [`/v3/sync` request][`v3`].
///
/// This returns a list of all the to-device events that were passed in but
/// encrypted ones were replaced with their decrypted version.
pub async fn from_sync_v2(
    response: &v3::Response,
    olm_machine: Option<&OlmMachine>,
) -> Result<Output> {
    process(
        olm_machine,
        response.to_device.events.clone(),
        &response.device_lists,
        &response.device_one_time_keys_count,
        response.device_unused_fallback_key_types.as_deref(),
        Some(response.next_batch.clone()),
    )
    .await
}

/// Process the to-device events and other related e2ee data.
///
/// This returns a list of all the to-device events that were passed in but
/// encrypted ones were replaced with their decrypted version.
async fn process(
    olm_machine: Option<&OlmMachine>,
    to_device_events: Vec<Raw<AnyToDeviceEvent>>,
    device_lists: &DeviceLists,
    one_time_keys_counts: &BTreeMap<OneTimeKeyAlgorithm, UInt>,
    unused_fallback_keys: Option<&[OneTimeKeyAlgorithm]>,
    next_batch_token: Option<String>,
) -> Result<Output> {
    let encryption_sync_changes = EncryptionSyncChanges {
        to_device_events,
        changed_devices: device_lists,
        one_time_keys_counts,
        unused_fallback_keys,
        next_batch_token,
    };

    Ok(if let Some(olm_machine) = olm_machine {
        // Let the crypto machine handle the sync response, this
        // decrypts to-device events, but leaves room events alone.
        // This makes sure that we have the decryption keys for the room
        // events at hand.
        let (events, room_key_updates) =
            olm_machine.receive_sync_changes(encryption_sync_changes).await?;

        let events = events
            .iter()
            // TODO: There is loss of information here, after calling `to_raw` it is not
            // possible to make the difference between a successfully decrypted event and a plain
            // text event. This information needs to be propagated to top layer at some point if
            // clients relies on custom encrypted to device events.
            .map(|p| p.to_raw())
            .collect();

        Output { decrypted_to_device_events: events, room_key_updates: Some(room_key_updates) }
    } else {
        // If we have no `OlmMachine`, just return the events that were passed in.
        // This should not happen unless we forget to set things up by calling
        // `Self::activate()`.
        Output {
            decrypted_to_device_events: encryption_sync_changes.to_device_events,
            room_key_updates: None,
        }
    })
}

pub struct Output {
    pub decrypted_to_device_events: Vec<Raw<AnyToDeviceEvent>>,
    pub room_key_updates: Option<Vec<RoomKeyInfo>>,
}
