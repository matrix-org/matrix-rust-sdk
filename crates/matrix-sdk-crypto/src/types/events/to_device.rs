// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::{collections::BTreeMap, fmt::Debug};

use ruma::{
    events::{
        dummy::ToDeviceDummyEvent,
        key::verification::{
            accept::ToDeviceKeyVerificationAcceptEvent, cancel::ToDeviceKeyVerificationCancelEvent,
            done::ToDeviceKeyVerificationDoneEvent, key::ToDeviceKeyVerificationKeyEvent,
            mac::ToDeviceKeyVerificationMacEvent, ready::ToDeviceKeyVerificationReadyEvent,
            request::ToDeviceKeyVerificationRequestEvent, start::ToDeviceKeyVerificationStartEvent,
        },
        room_key_request::ToDeviceRoomKeyRequestEvent,
        secret::request::{SecretName, ToDeviceSecretRequestEvent},
        EventContent, ToDeviceEventType,
    },
    serde::Raw,
    OwnedUserId, UserId,
};
use serde::{Deserialize, Serialize};
use serde_json::{
    value::{to_raw_value, RawValue},
    Value,
};
use zeroize::Zeroize;

use super::{
    forwarded_room_key::{ForwardedRoomKeyContent, ForwardedRoomKeyEvent},
    room::encrypted::EncryptedToDeviceEvent,
    room_key::RoomKeyEvent,
    secret_send::SecretSendEvent,
    EventType,
};
use crate::types::events::from_str;

/// An enum over the various to-device events we support.
#[derive(Debug)]
pub enum ToDeviceEvents {
    /// A to-device event of an unknown or custom type.
    Custom(ToDeviceCustomEvent),
    /// The `m.dummy` to-device event.
    Dummy(ToDeviceDummyEvent),

    /// The `m.key.verification.accept` to-device event.
    KeyVerificationAccept(ToDeviceKeyVerificationAcceptEvent),
    /// The `m.key.verification.cancel` to-device event.
    KeyVerificationCancel(ToDeviceKeyVerificationCancelEvent),
    /// The `m.key.verification.key` to-device event.
    KeyVerificationKey(ToDeviceKeyVerificationKeyEvent),
    /// The `m.key.verification.mac` to-device event.
    KeyVerificationMac(ToDeviceKeyVerificationMacEvent),
    /// The `m.key.verification.done` to-device event.
    KeyVerificationDone(ToDeviceKeyVerificationDoneEvent),
    /// The `m.key.verification.start` to-device event.
    KeyVerificationStart(ToDeviceKeyVerificationStartEvent),
    /// The `m.key.verification.ready` to-device event.
    KeyVerificationReady(ToDeviceKeyVerificationReadyEvent),
    /// The `m.key.verification.request` to-device event.
    KeyVerificationRequest(ToDeviceKeyVerificationRequestEvent),

    /// The `m.room.encrypted` to-device event.
    RoomEncrypted(EncryptedToDeviceEvent),
    /// The `m.room_key` to-device event.
    RoomKey(RoomKeyEvent),
    /// The `m.room_key_request` to-device event.
    RoomKeyRequest(ToDeviceRoomKeyRequestEvent),
    /// The `m.forwarded_room_key` to-device event.
    ForwardedRoomKey(Box<ForwardedRoomKeyEvent>),
    /// The `m.secret.send` to-device event.
    SecretSend(SecretSendEvent),
    /// The `m.secret.request` to-device event.
    SecretRequest(ToDeviceSecretRequestEvent),
}

impl ToDeviceEvents {
    /// The sender of the to-device event.
    pub fn sender(&self) -> &UserId {
        match self {
            ToDeviceEvents::Custom(e) => &e.sender,
            ToDeviceEvents::Dummy(e) => &e.sender,

            ToDeviceEvents::KeyVerificationAccept(e) => &e.sender,
            ToDeviceEvents::KeyVerificationCancel(e) => &e.sender,
            ToDeviceEvents::KeyVerificationKey(e) => &e.sender,
            ToDeviceEvents::KeyVerificationMac(e) => &e.sender,
            ToDeviceEvents::KeyVerificationDone(e) => &e.sender,
            ToDeviceEvents::KeyVerificationStart(e) => &e.sender,
            ToDeviceEvents::KeyVerificationReady(e) => &e.sender,
            ToDeviceEvents::KeyVerificationRequest(e) => &e.sender,

            ToDeviceEvents::RoomEncrypted(e) => &e.sender,
            ToDeviceEvents::RoomKey(e) => &e.sender,
            ToDeviceEvents::RoomKeyRequest(e) => &e.sender,
            ToDeviceEvents::ForwardedRoomKey(e) => &e.sender,

            ToDeviceEvents::SecretSend(e) => &e.sender,
            ToDeviceEvents::SecretRequest(e) => &e.sender,
        }
    }

    /// The event type of the to-device event.
    pub fn event_type(&self) -> ToDeviceEventType {
        match self {
            ToDeviceEvents::Custom(e) => ToDeviceEventType::from(e.event_type.to_owned()),
            ToDeviceEvents::Dummy(e) => e.content.event_type(),

            ToDeviceEvents::KeyVerificationAccept(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationCancel(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationKey(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationMac(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationDone(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationStart(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationReady(e) => e.content.event_type(),
            ToDeviceEvents::KeyVerificationRequest(e) => e.content.event_type(),

            ToDeviceEvents::RoomEncrypted(_) => ToDeviceEventType::RoomEncrypted,
            ToDeviceEvents::RoomKey(_) => ToDeviceEventType::RoomKey,
            ToDeviceEvents::RoomKeyRequest(e) => e.content.event_type(),
            ToDeviceEvents::ForwardedRoomKey(_) => ToDeviceEventType::ForwardedRoomKey,

            ToDeviceEvents::SecretSend(_) => ToDeviceEventType::SecretSend,
            ToDeviceEvents::SecretRequest(e) => e.content.event_type(),
        }
    }

    /// Serialize this event into a Raw variant while zeroizing any secrets it
    /// might contain.
    ///
    /// Secrets in Matrix are usually base64 encoded strings, zeroizing in this
    /// context means that the secret will be converted into an empty string.
    ///
    /// The following secrets will be zeroized by this method:
    ///
    /// * `m.room_key` - The `session_key` field.
    /// * `m.forwarded_room_key` - The `session_key` field.
    /// * `m.secret.send` - The `secret` field will be zeroized, unless the
    /// secret name of the matching `m.secret.request` event was
    /// `m.megolm_backup.v1`.
    ///
    /// **Warning**: Some events won't be able to be deserialized into the
    /// `ToDeviceEvents` type again since they might expect a valid `SessionKey`
    /// for `m.room.key` events or valid base64 for some other secrets.
    ///
    /// You can do a couple of things to avoid this problem:
    ///
    /// 1. Call `Raw::cast()` to convert the event to another, less strict type.
    /// [`AnyToDeviceEvent`] from Ruma will work.
    ///
    /// 2. Call `Raw::deserialize_as()` to deserialize into a less strict type.
    ///
    /// 3. Pass the event over FFI, losing the exact type information, this will
    /// mostl likely end up using a less strict type naturally.
    ///
    /// [`AnyToDeviceEvent`]: ruma::events::AnyToDeviceEvent
    pub(crate) fn serialize_zeroized(self) -> Result<Raw<ToDeviceEvents>, serde_json::Error> {
        let serialized = match self {
            ToDeviceEvents::Custom(_)
            | ToDeviceEvents::Dummy(_)
            | ToDeviceEvents::KeyVerificationAccept(_)
            | ToDeviceEvents::KeyVerificationCancel(_)
            | ToDeviceEvents::KeyVerificationKey(_)
            | ToDeviceEvents::KeyVerificationMac(_)
            | ToDeviceEvents::KeyVerificationDone(_)
            | ToDeviceEvents::KeyVerificationStart(_)
            | ToDeviceEvents::KeyVerificationReady(_)
            | ToDeviceEvents::KeyVerificationRequest(_)
            | ToDeviceEvents::RoomEncrypted(_)
            | ToDeviceEvents::RoomKeyRequest(_)
            | ToDeviceEvents::SecretRequest(_) => Raw::from_json(to_raw_value(&self)?),
            ToDeviceEvents::RoomKey(e) => {
                let event_type = e.content.event_type();
                let content = e.content.serialize_zeroized()?;

                #[derive(Serialize)]
                struct Helper<'a, C> {
                    sender: &'a UserId,
                    content: &'a Raw<C>,
                    #[serde(rename = "type")]
                    event_type: &'a str,
                }

                let helper = Helper { sender: &e.sender, content: &content, event_type };

                let raw_value = to_raw_value(&helper)?;

                Raw::from_json(raw_value)
            }
            ToDeviceEvents::ForwardedRoomKey(mut e) => {
                match &mut e.content {
                    ForwardedRoomKeyContent::MegolmV1AesSha2(c) => c.session_key.zeroize(),
                    #[cfg(feature = "experimental-algorithms")]
                    ForwardedRoomKeyContent::MegolmV2AesSha2(c) => c.session_key.zeroize(),
                    ForwardedRoomKeyContent::Unknown(_) => (),
                }

                Raw::from_json(to_raw_value(&e)?)
            }
            ToDeviceEvents::SecretSend(mut e) => {
                if let Some(SecretName::RecoveryKey) = e.content.secret_name {
                    // We don't zeroize the recovery key since it requires
                    // additional requests and possibly user-interaction to be
                    // verified. We let the user deal with this.
                } else {
                    e.content.secret.zeroize();
                }
                Raw::from_json(to_raw_value(&e)?)
            }
        };

        Ok(serialized)
    }
}

/// A to-device event with an unknown type and content.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ToDeviceCustomEvent {
    /// The sender of the to-device event.
    pub sender: OwnedUserId,
    /// The content of the to-device event.
    pub content: BTreeMap<String, Value>,
    /// The type of the to-device event.
    #[serde(rename = "type")]
    pub event_type: String,
    /// Any other unknown data of the to-device event.
    #[serde(flatten)]
    other: BTreeMap<String, Value>,
}

/// Generic to-device event with a known type and content.
#[derive(Debug, Deserialize)]
pub struct ToDeviceEvent<C>
where
    C: EventType + Debug + Sized + Serialize,
{
    /// The sender of the to-device event.
    pub sender: OwnedUserId,
    /// The content of the to-device event.
    pub content: C,
    /// Any other unknown data of the to-device event.
    #[serde(flatten)]
    pub(crate) other: BTreeMap<String, Value>,
}

impl<C> Serialize for ToDeviceEvent<C>
where
    C: EventType + Debug + Sized + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        #[derive(Serialize)]
        struct Helper<'a, C> {
            sender: &'a UserId,
            content: &'a C,
            #[serde(rename = "type")]
            event_type: &'a str,
            #[serde(flatten)]
            other: &'a BTreeMap<String, Value>,
        }

        let event_type = self.content.event_type();

        let helper =
            Helper { sender: &self.sender, content: &self.content, event_type, other: &self.other };

        helper.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for ToDeviceEvents {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        #[derive(Debug, Deserialize)]
        struct Helper<'a> {
            #[serde(rename = "type")]
            event_type: &'a str,
        }

        let json = Box::<RawValue>::deserialize(deserializer)?;
        let helper: Helper<'_> =
            serde_json::from_str(json.get()).map_err(serde::de::Error::custom)?;

        let json = json.get();

        Ok(match helper.event_type {
            "m.dummy" => ToDeviceEvents::Dummy(from_str(json)?),

            "m.key.verification.accept" => ToDeviceEvents::KeyVerificationAccept(from_str(json)?),
            "m.key.verification.cancel" => ToDeviceEvents::KeyVerificationCancel(from_str(json)?),
            "m.key.verification.done" => ToDeviceEvents::KeyVerificationDone(from_str(json)?),
            "m.key.verification.key" => ToDeviceEvents::KeyVerificationKey(from_str(json)?),
            "m.key.verification.mac" => ToDeviceEvents::KeyVerificationMac(from_str(json)?),
            "m.key.verification.start" => ToDeviceEvents::KeyVerificationStart(from_str(json)?),
            "m.key.verification.ready" => ToDeviceEvents::KeyVerificationReady(from_str(json)?),
            "m.key.verification.request" => ToDeviceEvents::KeyVerificationRequest(from_str(json)?),

            "m.room.encrypted" => ToDeviceEvents::RoomEncrypted(from_str(json)?),
            "m.room_key" => ToDeviceEvents::RoomKey(from_str(json)?),
            "m.forwarded_room_key" => ToDeviceEvents::ForwardedRoomKey(from_str(json)?),
            "m.room_key_request" => ToDeviceEvents::RoomKeyRequest(from_str(json)?),

            "m.secret.send" => ToDeviceEvents::SecretSend(from_str(json)?),
            "m.secret.request" => ToDeviceEvents::SecretRequest(from_str(json)?),

            _ => ToDeviceEvents::Custom(from_str(json)?),
        })
    }
}

impl Serialize for ToDeviceEvents {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            ToDeviceEvents::Custom(e) => e.serialize(serializer),
            ToDeviceEvents::Dummy(e) => e.serialize(serializer),

            ToDeviceEvents::KeyVerificationAccept(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationCancel(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationKey(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationMac(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationDone(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationStart(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationReady(e) => e.serialize(serializer),
            ToDeviceEvents::KeyVerificationRequest(e) => e.serialize(serializer),

            ToDeviceEvents::RoomEncrypted(e) => e.serialize(serializer),
            ToDeviceEvents::RoomKey(e) => e.serialize(serializer),
            ToDeviceEvents::RoomKeyRequest(e) => e.serialize(serializer),
            ToDeviceEvents::ForwardedRoomKey(e) => e.serialize(serializer),

            ToDeviceEvents::SecretSend(e) => e.serialize(serializer),
            ToDeviceEvents::SecretRequest(e) => e.serialize(serializer),
        }
    }
}

#[cfg(test)]
mod test {
    use matches::assert_matches;
    use serde_json::{json, Value};

    use super::ToDeviceEvents;

    fn custom_event() -> Value {
        json!({
            "sender": "@alice:example.org",
            "content": {
                "custom_key": "custom_value",
            },
            "m.custom.top": "something custom in the top",
            "type": "m.custom.event",
        })
    }

    fn key_verification_event() -> Value {
        json!({
            "sender": "@alice:example.org",
            "content": {
                "from_device": "AliceDevice2",
                "methods": [
                  "m.sas.v1"
                ],
                "timestamp": 1559598944869u64,
                "transaction_id": "S0meUniqueAndOpaqueString"
            },
            "type": "m.key.verification.request"
        })
    }

    fn dummy_event() -> Value {
        json!({
            "sender": "@alice:example.org",
            "content": {},
            "type": "m.dummy"
        })
    }

    fn secret_request_event() -> Value {
        json!({
            "sender": "@alice:example.org",
            "content": {
                "name": "org.example.some.secret",
                "action": "request",
                "requesting_device_id": "ABCDEFG",
                "request_id": "randomly_generated_id_9573"
            },
            "type": "m.secret.request"
        })
    }

    fn forwarded_room_key_event() -> Value {
        json!({
            "sender": "@alice:example.org",
            "content": {
            "algorithm": "m.megolm.v1.aes-sha2",
            "forwarding_curve25519_key_chain": [
                "hPQNcabIABgGnx3/ACv/jmMmiQHoeFfuLB17tzWp6Hw"
            ],
            "room_id": "!Cuyf34gef24t:localhost",
            "sender_claimed_ed25519_key": "aj40p+aw64yPIdsxoog8jhPu9i7l7NcFRecuOQblE3Y",
            "sender_key": "RF3s+E7RkTQTGF2d8Deol0FkQvgII2aJDf3/Jp5mxVU",
            "session_id": "X3lUlvLELLYxeTx4yOVu6UDpasGEVO0Jbu+QFnm0cKQ",
            "session_key": "AQAAAAq2JpkMceK5f6JrZPJWwzQTn59zliuIv0F7apVLXDcZCCT\
                            3LqBjD21sULYEO5YTKdpMVhi9i6ZSZhdvZvp//tzRpDT7wpWVWI\
                            00Y3EPEjmpm/HfZ4MMAKpk+tzJVuuvfAcHBZgpnxBGzYOc/DAqa\
                            pK7Tk3t3QJ1UMSD94HfAqlb1JF5QBPwoh0fOvD8pJdanB8zxz05\
                            tKFdR73/vo2Q/zE3"
            },
            "type": "m.forwarded_room_key"
        })
    }

    fn room_key_request_event() -> Value {
        json!({
            "sender": "@alice:example.org",
            "content": {
                "action": "request",
                "body": {
                  "algorithm": "m.megolm.v1.aes-sha2",
                  "room_id": "!Cuyf34gef24t:localhost",
                  "sender_key": "RF3s+E7RkTQTGF2d8Deol0FkQvgII2aJDf3/Jp5mxVU",
                  "session_id": "X3lUlvLELLYxeTx4yOVu6UDpasGEVO0Jbu+QFnm0cKQ"
                },
                "request_id": "1495474790150.19",
                "requesting_device_id": "RJYKSTBOIE"
            },
            "type": "m.room_key_request"
        })
    }

    #[test]
    fn deserialization() -> Result<(), serde_json::Error> {
        macro_rules! assert_serialization_roundtrip {
            ( $( $json:path => $to_device_events:ident ),* $(,)? ) => {
                $(
                    let json = $json();
                    let event: ToDeviceEvents = serde_json::from_value(json.clone())?;

                    assert_matches!(event, ToDeviceEvents::$to_device_events(_));
                    let serialized = serde_json::to_value(event)?;
                    assert_eq!(json, serialized);
                )*
            }
        }

        assert_serialization_roundtrip!(
            // `m.room_key
            crate::types::events::room_key::test::json => RoomKey,

            // `m.forwarded_room_key`
            forwarded_room_key_event => ForwardedRoomKey,

            // `m.room_key_request`
            room_key_request_event => RoomKeyRequest,

            // `m.secret.send`
            crate::types::events::secret_send::test::json => SecretSend,

            // `m.secret.request`
            secret_request_event => SecretRequest,

            // Unknown event
            custom_event => Custom,

            // `m.key.verification.request`
            key_verification_event => KeyVerificationRequest,

            // `m.dummy`
            dummy_event => Dummy,

            // `m.room.encrypted`
            crate::types::events::room::encrypted::test::to_device_json => RoomEncrypted,
        );

        Ok(())
    }
}
