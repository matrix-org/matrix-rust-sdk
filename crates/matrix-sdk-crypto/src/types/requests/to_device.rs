// Copyright 2020 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::{collections::BTreeMap, iter};

use ruma::{
    events::{AnyToDeviceEventContent, EventContent, ToDeviceEventType},
    serde::Raw,
    to_device::DeviceIdOrAllDevices,
    OwnedDeviceId, OwnedTransactionId, OwnedUserId, TransactionId, UserId,
};
use serde::{Deserialize, Serialize};

/// Customized version of
/// `ruma_client_api::to_device::send_event_to_device::v3::Request`
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ToDeviceRequest {
    /// Type of event being sent to each device.
    pub event_type: ToDeviceEventType,

    /// A request identifier unique to the access token used to send the
    /// request.
    pub txn_id: OwnedTransactionId,

    /// A map of users to devices to a content for a message event to be
    /// sent to the user's device. Individual message events can be sent
    /// to devices, but all events must be of the same type.
    /// The content's type for this field will be updated in a future
    /// release, until then you can create a value using
    /// `serde_json::value::to_raw_value`.
    pub messages:
        BTreeMap<OwnedUserId, BTreeMap<DeviceIdOrAllDevices, Raw<AnyToDeviceEventContent>>>,
}

impl ToDeviceRequest {
    /// Create a new owned to-device request
    ///
    /// # Arguments
    ///
    /// * `recipient` - The ID of the user that should receive this to-device
    ///   event.
    ///
    /// * `recipient_device` - The device that should receive this to-device
    ///   event, or all devices.
    ///
    /// * `event_type` - The type of the event content that is getting sent out.
    ///
    /// * `content` - The content of the to-device event.
    pub fn new(
        recipient: &UserId,
        recipient_device: impl Into<DeviceIdOrAllDevices>,
        event_type: &str,
        content: Raw<AnyToDeviceEventContent>,
    ) -> Self {
        let event_type = ToDeviceEventType::from(event_type);
        let user_messages = iter::once((recipient_device.into(), content)).collect();
        let messages = iter::once((recipient.to_owned(), user_messages)).collect();

        ToDeviceRequest { event_type, txn_id: TransactionId::new(), messages }
    }

    pub(crate) fn for_recipients(
        recipient: &UserId,
        recipient_devices: Vec<OwnedDeviceId>,
        content: &AnyToDeviceEventContent,
        txn_id: OwnedTransactionId,
    ) -> Self {
        let event_type = content.event_type();
        let raw_content = Raw::new(content).expect("Failed to serialize to-device event");

        if recipient_devices.is_empty() {
            Self::new(
                recipient,
                DeviceIdOrAllDevices::AllDevices,
                &event_type.to_string(),
                raw_content,
            )
        } else {
            let device_messages = recipient_devices
                .into_iter()
                .map(|d| (DeviceIdOrAllDevices::DeviceId(d), raw_content.clone()))
                .collect();

            let messages = iter::once((recipient.to_owned(), device_messages)).collect();

            ToDeviceRequest { event_type, txn_id, messages }
        }
    }

    pub(crate) fn with_id_raw(
        recipient: &UserId,
        recipient_device: impl Into<DeviceIdOrAllDevices>,
        content: Raw<AnyToDeviceEventContent>,
        event_type: ToDeviceEventType,
        txn_id: OwnedTransactionId,
    ) -> Self {
        let user_messages = iter::once((recipient_device.into(), content)).collect();
        let messages = iter::once((recipient.to_owned(), user_messages)).collect();

        ToDeviceRequest { event_type, txn_id, messages }
    }

    pub(crate) fn with_id(
        recipient: &UserId,
        recipient_device: impl Into<DeviceIdOrAllDevices>,
        content: &AnyToDeviceEventContent,
        txn_id: OwnedTransactionId,
    ) -> Self {
        let event_type = content.event_type();
        let raw_content = Raw::new(content).expect("Failed to serialize to-device event");

        let user_messages = iter::once((recipient_device.into(), raw_content)).collect();
        let messages = iter::once((recipient.to_owned(), user_messages)).collect();

        ToDeviceRequest { event_type, txn_id, messages }
    }

    /// Get the number of unique messages this request contains.
    ///
    /// *Note*: A single message may be sent to multiple devices, so this may or
    /// may not be the number of devices that will receive the messages as well.
    pub fn message_count(&self) -> usize {
        self.messages.values().map(|d| d.len()).sum()
    }
}
