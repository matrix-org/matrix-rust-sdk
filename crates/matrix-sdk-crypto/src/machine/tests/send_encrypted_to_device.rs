// Copyright 2020 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use assert_matches2::assert_matches;
use matrix_sdk_test::async_test;
use ruma::to_device::DeviceIdOrAllDevices;
use serde_json::{json, value::to_raw_value};

use crate::{
    machine::{
        test_helpers::{get_machine_pair, get_machine_pair_with_session},
        tests,
    },
    types::events::ToDeviceEvent,
    utilities::json_convert,
    EncryptionSyncChanges, OlmError, ToDeviceRequest,
};

#[async_test]
async fn test_send_encrypted_to_device() {
    let (alice, bob) =
        get_machine_pair_with_session(tests::alice_id(), tests::user_id(), false).await;

    let custom_event_type = "m.new_device";

    let custom_content = json!({
            "device_id": "XYZABCDE",
            "rooms": ["!726s6s6q:example.com"]
    });

    let device = alice.get_device(bob.user_id(), bob.device_id(), None).await.unwrap().unwrap();
    let raw_encrypted = device
        .encrypt_event_raw(custom_event_type, &custom_content)
        .await
        .expect("Should have encryted the content");

    let request = ToDeviceRequest::new(
        bob.user_id(),
        DeviceIdOrAllDevices::DeviceId(tests::bob_device_id().to_owned()),
        "m.room.encrypted",
        raw_encrypted.cast(),
    );

    assert_eq!("m.room.encrypted", request.event_type.to_string());

    let messages = &request.messages;
    assert_eq!(1, messages.len());
    assert!(messages.get(bob.user_id()).is_some());
    let target_devices = messages.get(bob.user_id()).unwrap();
    assert_eq!(1, target_devices.len());
    assert!(target_devices
        .get(&DeviceIdOrAllDevices::DeviceId(tests::bob_device_id().to_owned()))
        .is_some());

    let event = ToDeviceEvent::new(
        alice.user_id().to_owned(),
        tests::to_device_requests_to_content(vec![request.clone().into()]),
    );

    let event = json_convert(&event).unwrap();

    let sync_changes = EncryptionSyncChanges {
        to_device_events: vec![event],
        changed_devices: &Default::default(),
        one_time_keys_counts: &Default::default(),
        unused_fallback_keys: None,
        next_batch_token: None,
    };

    let (decrypted, _) = bob.receive_sync_changes(sync_changes).await.unwrap();

    assert_eq!(1, decrypted.len());

    let decrypted_event = decrypted[0].deserialize().unwrap();

    assert_eq!(decrypted_event.event_type().to_string(), custom_event_type.to_owned());

    let decrypted_value = to_raw_value(&decrypted[0]).unwrap();
    let decrypted_value = serde_json::to_value(decrypted_value).unwrap();

    assert_eq!(
        decrypted_value.get("content").unwrap().get("device_id").unwrap().as_str().unwrap(),
        custom_content.get("device_id").unwrap().as_str().unwrap(),
    );

    assert_eq!(
        decrypted_value.get("content").unwrap().get("rooms").unwrap().as_array().unwrap(),
        custom_content.get("rooms").unwrap().as_array().unwrap(),
    );
}

#[async_test]
async fn test_send_encrypted_to_device_no_session() {
    let (alice, bob, _) = get_machine_pair(tests::alice_id(), tests::user_id(), false).await;

    let custom_event_type = "m.new_device";

    let custom_content = json!({
            "device_id": "XYZABCDE",
            "rooms": ["!726s6s6q:example.com"]
    });

    let encryption_result = alice
        .get_device(bob.user_id(), tests::bob_device_id(), None)
        .await
        .unwrap()
        .unwrap()
        .encrypt_event_raw(custom_event_type, &custom_content)
        .await;

    assert_matches!(encryption_result, Err(OlmError::MissingSession));
}
