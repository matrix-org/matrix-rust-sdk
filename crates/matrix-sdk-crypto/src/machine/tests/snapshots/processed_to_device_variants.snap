---
source: crates/matrix-sdk-crypto/src/machine/tests/send_encrypted_to_device.rs
expression: "processed_event.to_raw().deserialize_as::<Value>().unwrap()"
---
{
  "content": {
    "call_id": "",
    "device_id": "XYZABCDE",
    "keys": [
      {
        "index": 0,
        "key": "I7qrSrCR7Yo9B4iGVnR8IA"
      }
    ]
  },
  "keys": {
    "ed25519": "[sender_ed25519_key]"
  },
  "recipient": "@bob:example.com",
  "recipient_keys": {
    "ed25519": "[recipient_sender_key]"
  },
  "sender": "@alice:example.org",
  "sender_device_keys": {
    "algorithms": "++REDACTED++",
    "device_id": "[ABCDEFGH]",
    "keys": "++REDACTED++",
    "signatures": "++REDACTED++",
    "user_id": "@alice:example.org"
  },
  "type": "rtc.call.encryption_keys"
}
